extension-androidauto/src/main/java/com/mapbox/maps/extension/androidauto/BaseWidgetRenderer.kt

```diff
@@ -0,0 +1,393 @@
+      GLES20.glDetachShader(program, fragmentShader)
+      GLES20.glDeleteShader(vertexShader)
+      GLES20.glDeleteShader(fragmentShader)
+      GLES20.glDeleteTextures(textures.size, textures, 0)
+      GLES20.glDeleteProgram(program)
+      program = 0
+    }
+    dirty = true
+  }
+
+  private fun createTextureFromBitmap() {
+    if (!bitmap.isRecycled) {
+      // generate texture
+      GLES20.glGenTextures(1, textures, 0)
+      // bind texture
+      GLES20.glBindTexture(GLES20.GL_TEXTURE_2D, textures[0])
+      // set the color filter is reduced pixel color closest to the coordinates of a pixel in texture drawn as required
+      GLES20.glTexParameterf(
+        GLES20.GL_TEXTURE_2D,
+        GLES20.GL_TEXTURE_MIN_FILTER,
+        GLES20.GL_NEAREST.toFloat()
+      )
+      // set the amplification filter using texture coordinates to the nearest number of colors, obtained by the weighted averaging algorithm requires pixel color drawn
+      GLES20.glTexParameterf(
+        GLES20.GL_TEXTURE_2D,
+        GLES20.GL_TEXTURE_MAG_FILTER,
+        GLES20.GL_LINEAR.toFloat()
+      )
+      // Set the circumferential direction S, texture coordinates taken to [1 / 2n, 1-1 / 2n]. Will never lead to integration and border
+      GLES20.glTexParameterf(
+        GLES20.GL_TEXTURE_2D,
+        GLES20.GL_TEXTURE_WRAP_S,
+        GLES20.GL_CLAMP_TO_EDGE.toFloat()
+      )
+      // Set the circumferential direction T, taken to texture coordinates [1 / 2n, 1-1 / 2n]. Will never lead to integration and border
+      GLES20.glTexParameterf(
+        GLES20.GL_TEXTURE_2D,
+        GLES20.GL_TEXTURE_WRAP_T,
+        GLES20.GL_CLAMP_TO_EDGE.toFloat()
+      )
+      // The parameters specified above, generates a 2D texture
+      GLUtils.texImage2D(GLES20.GL_TEXTURE_2D, 0, bitmap, 0)
+      bitmap.recycle()
+    }
+  }
+
+  fun updateBitmap(bitmap: Bitmap) {
+    this.bitmap = bitmap
+    onSizeChanged(width, height)
+    dirty = true
+  }
+
+  fun rotate(bearing: Float) {
+    Matrix.setIdentityM(rotationMatrixData, 0)
+    Matrix.rotateM(rotationMatrixData, 0, bearing, 0f, 0f, 1f)
+    rotationDirty = true
+    dirty = true
+  }
+
+  // TODO fun translate() needed?
+
+  companion object {
+    private const val TAG = "BaseWidgetRenderer"
+
+    // number of coordinates per vertex in this array
+    private const val COORDS_PER_VERTEX = 2
+    private const val BYTES_PER_FLOAT = 4
+    private const val VERTEX_STRIDE = COORDS_PER_VERTEX * BYTES_PER_FLOAT // 4 bytes per vertex
+    private const val VERTEX_COUNT = 4  // 4 vertex in total
+
+    private val VERTEX_SHADER_CODE = """
+      uniform mat4 uScreen;
+      uniform mat4 uTranslate;
+      uniform mat4 uRotation;
+      attribute vec2 vPosition;
+      attribute vec2 vCoordinate;
+      varying vec2 aCoordinate;
+      void main() {
+        aCoordinate = vCoordinate;
+        gl_Position = uScreen * uTranslate  * uRotation * vec4(vPosition, 0.0, 1.0);
```

Comment:
Can we actually use https://developer.android.com/reference/android/graphics/Matrix class to calculate single MVP matrix instead of passing so many `uniform mat4`?

Reply:
Done